py/gc.c

```diff
@@ -203,29 +203,13 @@ bool gc_is_locked(void) {
 #endif
 #endif
 
-// ptr should be of type void*
-#define VERIFY_MARK_AND_PUSH(ptr) \
-    do { \
-        if (VERIFY_PTR(ptr)) { \
-            size_t _block = BLOCK_FROM_PTR(ptr); \
-            if (ATB_GET_KIND(_block) == AT_HEAD) { \
-                /* an unmarked head, mark it, and push it on gc stack */ \
-                TRACE_MARK(_block, ptr); \
-                ATB_HEAD_TO_MARK(_block); \
-                if (MP_STATE_MEM(gc_sp) < &MP_STATE_MEM(gc_stack)[MICROPY_ALLOC_GC_STACK_SIZE]) { \
-                    *MP_STATE_MEM(gc_sp)++ = _block; \
-                } else { \
-                    MP_STATE_MEM(gc_stack_overflow) = 1; \
-                } \
-            } \
-        } \
-    } while (0)
-
-STATIC void gc_drain_stack(void) {
-    while (MP_STATE_MEM(gc_sp) > MP_STATE_MEM(gc_stack)) {
-        // pop the next block off the stack
-        size_t block = *--MP_STATE_MEM(gc_sp);
-
+// Take the given block as the topmost block on the stack. Check all it's
+// children: mark the unmarked child blocks and put those newly marked
+// blocks on the stack. When all children have been checked, pop off the
+// topmost block on the stack and repeat with that one.
+STATIC void gc_drain_stack(size_t block) {
```

Comment:
Maybe this function needs to be renamed - it isn't really draining the stack anymore as no one outside `gc_drain_stack` pushes anything to the stack.
Maybe something like `gc_mark_children_recursive`? It isn't recursive in the common way (calling itself) but it implements a sort of recursion by using it's own stack, hence the name.

Reply:
Yes it makes sense to rename it.  But it would be good to stay away from the word "recursive" if possible because embedded stuff is ideally not recursive :)  Something like `gc_mark_subtree` or `gc_mark_descendants`.

Reply:
I've renamed it to `gc_mark_subtree`.